docs: Update NVIDIA GPU passthrough documentation

manuelh-dev opened the pull request:
Amend the existing documentation with structural improvements and pointers to a reference implementation for a Kata Containers stack for NVIDIA GPUs. The new page on the reference implementation provides the following:
- components involved in the NVIDIA (TEE) GPU scenario
- orchestration flow for GPU passthrough scenarios
- deployment guidance

dcmiddle reviewed (approved):
This is great documentation. It answered a lot of questions for me I couldn't find elsewhere.
I've reviewed it in detail based on my knowledge of CoCo. I don't have any corrections / suggestions.

Apokleos reviewed (commented):
Thx @manuelh-dev comments left and FYI

LandonTClipp commented on docs/use-cases/NVIDIA-GPU-passthrough-and-Kata.md:
```diff
@@ -1,621 +1,346 @@
+Containerd in turn invokes the Kata runtime with the OCI spec.
+
+### Kata runtime
+The Kata runtime receives the OCI spec with these annotations and then queries kubelet's Pod Resources API using the information from the annotations to acquire the GPU device ID allocated by the sandbox device plugin.
```
I think you should note that the Pod Resource API is a new feature that requires explicit enablement for older k8s versions. And also perhaps specify that this is in the cold-plugging case. The hot-plug case is quite different.

LandonTClipp commented on docs/use-cases/NVIDIA-GPU-passthrough-and-Kata.md:
```diff
@@ -1,621 +1,346 @@
+The runtime then consults the host CDI specifications and determines the device path the GPU is backed by (e.g., `/dev/vfio/devices/vfio0`).
+Then, the runtime resolves the device's PCI BDF (e.g., `0000:01:00.0`).
+
+For NVIDIA GPU TEE workloads, the runtime cold-plugs the GPU by launching QEMU with relevant parameters for device passthrough (e.g., `-device vfio-pci,host=0000:01:00.0,id=gpu0,bus=pcie.0,addr=0x2`).
```
Another thing: this should be using IOMMUFD if using the Go runtime because your example passes in the IOMMUFD cdev. 

That could also be too much information just for an overview. Just an idea.

LandonTClipp commented on docs/use-cases/NVIDIA-GPU-passthrough-and-Kata.md:
```diff
@@ -1,621 +1,346 @@
+Then, the runtime resolves the device's PCI BDF (e.g., `0000:01:00.0`).
+
+For NVIDIA GPU TEE workloads, the runtime cold-plugs the GPU by launching QEMU with relevant parameters for device passthrough (e.g., `-device vfio-pci,host=0000:01:00.0,id=gpu0,bus=pcie.0,addr=0x2`).
+The runtime then creates "inner runtime" CDI annotations for workload containers which are sent to kata-agent as part of the create container request.
```
I don't think it's the containerd shim that creates the inner runtime, it's the kata agent that asks the nvidia-ctk for it. Someone should double check me here. The agent receives the device IDs from the shim but not enough information on where the devices are exposed in the VM (again, someone correct me if I'm wrong, I could be).

manuelh-dev replied:
I am trying to clarify this right now - please take a look at `createCDIAnnotation` in `src/runtime/virtcontainers/container.go` - this is what I was referring to here. 

I think, in your comment you were referring to the inner CDI specification file

LandonTClipp commented on docs/use-cases/NVIDIA-GPU-passthrough-and-Kata.md:
````diff
@@ -1,621 +1,346 @@
+kubectl logs -n default cuda-vectoradd-kata
 ```
 
-## Build Kata Containers kernel with GPU support
````
I think it should be mentioned that QEMU is really the best-supported VMM for GPUs. I initially tried using cloud-hypervisor and realized after much wasted time that it wasn't fully supported for GPU passthrough. The docs should at least explicitly call this out, especially being the "GPU passthrough" docs.

The deleted docs here are also misleading when it states you have to build a kernel with GPU support. That is not true because Kata packages the NVIDIA GPU runtime. I also went down a wrong path because of what these docs say.

Apokleos replied:
> I think it should be mentioned that QEMU is really the best-supported VMM for GPUs. I initially tried using cloud-hypervisor and realized after much wasted time that it wasn't fully supported for GPU passthrough. The docs should at least explicitly call this out, especially being the "GPU passthrough" docs.
> 
Thx for your feedback, and agree that the doc is indeed partially out of date, and it's why @manuelh-dev do such work. By the way, could please figure out which one brought you such bad experience, runtime-go/clh or runtime-rs/cloud-hypervisor ? we can disscuss about it if we can do something to support it.
>
> The deleted docs here are also misleading when it states you have to build a kernel with GPU support. That is not true because Kata packages the NVIDIA GPU runtime. I also went down a wrong path because of what these docs say.
>
yeah, kata packages the NVIDIA GPU runtime is most for runtime-go/qemu

Apokleos replied:
> I used the Go runtime for everything. Rust runtime seems not in feature parity with go for now. 
>
Do you mean the GPU feature parity ?
>
>So I tried the go runtime with clh. Do you have a different experience with the Rust runtime with GPU >passthrough? I know the Ant group was the one that introduced rust so I'd like to hear more your >experience.
>
Make sense for the experience. Actually runtime-rs/cloud-hypervisor supports GPU passthrough well, and I have ever done some work on this combination of runtime-rs and cloud-hypervisor, but we should know the reality that cloud-hypervisor is a lightweight vmm which cannot support such complex and poweful abilities as qemu vmm.
And builtin lightweight vmm called Dragonball also support GPU passthrough.

manuelh-dev commented on docs/use-cases/NVIDIA-GPU-passthrough-and-Kata.md:
```diff
@@ -1,621 +1,346 @@
+The runtime then creates "inner runtime" CDI annotations for workload containers which are sent to kata-agent as part of the create container request.
+This provides kata-agent with the relevant information into which container to map the passthrough GPU.
+
+For non-TEE NVIDIA GPU workloads, the hot-plug method is being used - with the difference being the GPU being attached to the pod VM with QEMU before container creation, but after the VM was created (using a QMP `device_add` command).
```
@zvonkok - this is on our road map, is this correct?

manuelh-dev committed:
docs: Improve structure and flow of NVIDIA guide

- Apply a few structural/grouping changes and improve flow
- Group build sections together
- Move usage examples to last section

Signed-off-by: Manuel Huber <manuelh@nvidia.com>

manuelh-dev committed:
docs: Create NVIDIA GPU passthrough QEMU scenario

Create a new page for a reference implementation for Kubernetes
using QEMU, the go shim and an NVIDIA rootfs. The new page
contains information on:
- components involved in the NVIDIA (TEE) GPU scenario
- orchestration flow for GPU passthrough scenarios
- deployment guidance

Signed-off-by: Manuel Huber <manuelh@nvidia.com>

zvonkok commented:
For GPU support QEMU is THE reference platform. We're working on Dragonball support to some extent but we're excluding any other VMM.

Apokleos reviewed (commented):
Thx @manuelh-dev for this document. Honestly, I don't think it's only a document for NVIDIA GPU passthrough for qemu, but a technical internal for us to know more detailed implementations of the entire workflow for GPU passthrough with CDI from kubernetes to kata-runtime/kata-agent. 
I walk through the document and have some unclear parts with comments left.

manuelh-dev commented on docs/use-cases/NVIDIA-GPU-passthrough-and-Kata-QEMU.md:
```diff
@@ -0,0 +1,544 @@
+boots and mounts the root filesystem. After this, NVRC starts as the initial
+process, scans for NVIDIA GPUs on the PCI bus, loads the NVIDIA kernel
+modules, waits for driver initialization, creates the device nodes, and
+initializes the GPU hardware (using the `nvidia-smi` binary). NVRC also
```
Very good point - I have re-worked the section `#### UVM orchestration flow`, differentiating explicitly between hot- and cold-plug

manuelh-dev commented on docs/use-cases/NVIDIA-GPU-passthrough-and-Kata-QEMU.md:
```diff
@@ -0,0 +1,544 @@
+
+When the Kata runtime sends the create container request, the Kata agent
+parses the "inner runtime" CDI annotation. For example, given the annotation
+`"cdi.k8s.io/vfio1": "nvidia.com/gpu=0"`, the agent looks up device `0` in
```
I am resolving this comment - a similar comment on clarification of inner and outer runtime CDI annotation already exists above :)

Apokleos reviewed (commented):
Thx @manuelh-dev Still one comment left to make it clear.

Apokleos reviewed (approved):
Thx a lot @manuelh-dev for your effort and patience. The document deeply clarify the workflow of GPU coldplug and hotplug. And I believe it will help more users to know about how GPUs work in kata-containers.
LGTM !

manuelh-dev commented on docs/use-cases/NVIDIA-GPU-passthrough-and-Kata-QEMU.md:
```diff
@@ -0,0 +1,590 @@
+
+> **Note:**
+>
+> We recommend to configure your Kubelet with a higher
```
Good point, I had remarked this a bit further above in the previous version, but this was indeed very vague:
```
First, set up your Kubernetes cluster. For instance, in Kata CI, our NVIDIA
jobs use a single-node vanilla Kubernetes cluster with a 2.x containerd version and Kata's current supported Kubernetes version (in which the Pod Resources API is enabled by default - for older Kubernetes versions, this API may require explicit enablement
```


I made this more clear by adding a separate "Note", just below your comment in my updated revision:
```
> **Note:**
>
> We recommend to configure your Kubelet with a higher
> `runtimeRequestTimeout` timeout value than the two minute default timeout.
> Using the guest-pull mechanism, pulling large images may take a significant
> amount of time and may delay container start, possibly leading your Kubelet
> to de-allocate your pod before it transitions from the *container created*
> to the *container running* state.
```

 Thanks a lot!

zvonkok reviewed (approved):
LGTM, thanks for the effort on this one @manuelh-dev
